File: logiQ/backend/models/question.go
package models

import "go.mongodb.org/mongo-driver/bson/primitive"

type QuestionType string

const (
	QuestionTypeSingleChoice   QuestionType = "singleChoice"
	QuestionTypeMultipleChoice QuestionType = "multipleChoice"
	QuestionTypeTrueFalse      QuestionType = "trueFalse"
)

type QuestionCategory string

const (
	QuestionCategoryTruthTable  QuestionCategory = "truthTable"
	QuestionCategoryEquivalence QuestionCategory = "equivalence"
	QuestionCategoryInference   QuestionCategory = "inference"
)

type QuestionDifficulty string

const (
	QuestionDifficultyEasy   QuestionDifficulty = "easy"
	QuestionDifficultyMedium QuestionDifficulty = "medium"
	QuestionDifficultyHard   QuestionDifficulty = "hard"
)

type Question struct {
	ID                 primitive.ObjectID `json:"_id,omitempty" bson:"_id,omitempty"`
	QuestionText       string             `json:"question_text" bson:"question_text" binding:"required"`
	Options            []string           `json:"options" bson:"options" binding:"required"`
	CorrectAnswerIndex []int              `json:"correct_answer_index" bson:"correct_answer_index" binding:"required"`
	Type               QuestionType       `json:"type" bson:"type" binding:"required,oneof=singleChoice multipleChoice trueFalse"`    // "singleChoice" | "multipleChoice" | "trueFalse"
	Category           QuestionCategory   `json:"category" bson:"category" binding:"required,oneof=truthTable equivalence inference"` // "truthTable" | "equivalence" | "inference"
	Difficulty         QuestionDifficulty `json:"difficulty" bson:"difficulty" binding:"required,oneof=easy medium hard"`             // "easy" | "medium" | "hard"
	IsActive           bool               `json:"is_active" bson:"is_active"`
}

type GenerateQuestionRequest struct {
	Number int `json:"number" bson:"number" binding:"required"`
	// 以下三个字段可选，不提供则表示不限制
	Category   QuestionCategory   `json:"category" bson:"category" binding:"omitempty,oneof=truthTable equivalence inference"`
	Difficulty QuestionDifficulty `json:"difficulty" bson:"difficulty" binding:"omitempty,oneof=easy medium hard"`
	Type       QuestionType       `json:"type" bson:"type" binding:"omitempty,oneof=singleChoice multipleChoice trueFalse"`
}

type GetQuestionListRequest struct {
	Category   QuestionCategory   `json:"category,omitempty" form:"category" bson:"category,omitempty" binding:"omitempty,oneof=truthTable equivalence inference"`
	Difficulty QuestionDifficulty `json:"difficulty,omitempty" form:"difficulty" bson:"difficulty,omitempty" binding:"omitempty,oneof=easy medium hard"`
	Type       QuestionType       `json:"type,omitempty" form:"type" bson:"type,omitempty" binding:"omitempty,oneof=singleChoice multipleChoice trueFalse"`
	Page       int                `json:"page,omitempty" form:"page" bson:"page,omitempty" binding:"omitempty,min=1"`
	PageSize   int                `json:"page_size,omitempty" form:"page_size" bson:"page_size,omitempty" binding:"omitempty,min=1,max=100"`
}

// QuestionResponseForAdmin 用于管理员查看的题目详情，包含正确率
type QuestionResponseForAdmin struct {
	ID                 primitive.ObjectID `json:"_id,omitempty" bson:"_id,omitempty"`
	QuestionText       string             `json:"question_text" bson:"question_text" binding:"required"`
	Options            []string           `json:"options" bson:"options" binding:"required"`
	CorrectAnswerIndex []int              `json:"correct_answer_index" bson:"correct_answer_index" binding:"required"`
	Type               QuestionType       `json:"type" bson:"type" binding:"required,oneof=singleChoice multipleChoice trueFalse"`    // "singleChoice" | "multipleChoice" | "trueFalse"
	Category           QuestionCategory   `json:"category" bson:"category" binding:"required,oneof=truthTable equivalence inference"` // "truthTable" | "equivalence" | "inference"
	Difficulty         QuestionDifficulty `json:"difficulty" bson:"difficulty" binding:"required,oneof=easy medium hard"`             // "easy" | "medium" | "hard"
	IsActive           bool               `json:"is_active" bson:"is_active"`

	// 新增的统计字段
	TotalAnswers   int64   `json:"total_answers" bson:"total_answers"`
	CorrectAnswers int64   `json:"correct_answers" bson:"correct_answers"`
	AccuracyRate   float64 `json:"accuracy_rate" bson:"accuracy_rate"`
}
type BatchDeleteRequest struct {
	IDs []string `json:"ids" binding:"required,dive,hexadecimal"`
}

File: logiQ/frontend/lib/models/question.dart

class Question {

  final String id;
  final String questionText;
  final List<String> options;
  final List<int> correctAnswerIndex;

  final QuestionType type;
  final QuestionCategory category;
  final QuestionDifficulty difficulty;

  Question({
    required this.id,
    required this.questionText,
    required this.options,
    required this.correctAnswerIndex,
    required this.type,
    required this.category,
    required this.difficulty,
  });

  Question copyWith({
    String? id,
    String? questionText,
    List<String>? options,
    List<int>? correctAnswerIndex,
    QuestionType? type,
    QuestionCategory? category,
    QuestionDifficulty? difficulty,
  }) {
    return Question(
      id: id ?? this.id,
      questionText: questionText ?? this.questionText,
      options: options ?? this.options,
      correctAnswerIndex: correctAnswerIndex ?? this.correctAnswerIndex,
      type: type ?? this.type,
      category: category ?? this.category,
      difficulty: difficulty ?? this.difficulty,
    );
  }

  // JSON 序列化支持
  factory Question.fromJson(Map<String, dynamic> json) => Question(
    id: json['_id'] ?? json['id'] ?? '', // 兼容MongoDB的_id字段
    questionText: json['question_text'] ?? '',
    options: List<String>.from(json['options'] ?? []),
    correctAnswerIndex: List<int>.from(json['correct_answer_index'] ?? []),
    type: QuestionType.values.firstWhere(
      (e) => e.name == json['type'],
      orElse: () => QuestionType.singleChoice,
    ),
    category: QuestionCategory.values.firstWhere(
      (e) => e.name == json['category'],
      orElse: () => QuestionCategory.truthTable,
    ),
    difficulty: QuestionDifficulty.values.firstWhere(
      (e) => e.name == json['difficulty'],
      orElse: () => QuestionDifficulty.easy,
    ),
  );

  Map<String, dynamic> toJson() => {
    '_id': id,  // 兼容MongoDB的_id字段
    'question_text': questionText,
    'options': options,
    'correct_answer_index': correctAnswerIndex,
    'type': type.name,
    'category': category.name,
    'difficulty': difficulty.name,
  };

}

enum QuestionType {
  singleChoice,
  multipleChoice,
  trueFalse,
}

extension QuestionTypeExtension on QuestionType {
  String get displayName {
    switch (this) {
      case QuestionType.singleChoice:
        return 'Single Choice';
      case QuestionType.multipleChoice:
        return 'Multiple Choice';
      case QuestionType.trueFalse:
        return 'True / False';
    }
  }

  static String getDisplayNameFromName(String name) {
    try {
      final type = QuestionType.values.firstWhere((e) => e.name == name);
      return type.displayName;
    } catch (e) {
      return name; // 如果找不到匹配的枚举值，返回原始名称
    }
  }

  static String getNameFromDisplayName(String displayName) {
    try {
      final type = QuestionType.values.firstWhere((e) => e.displayName == displayName);
      return type.name;
    } catch (e) {
      return displayName; // 如果找不到匹配的枚举值，返回原始显示名称
    }
  }
}

enum QuestionCategory {
  truthTable,
  equivalence,
  inference,
}

extension QuestionCategoryExtension on QuestionCategory {
  String get displayName {
    switch (this) {
      case QuestionCategory.truthTable:
        return 'Truth Table';
      case QuestionCategory.equivalence:
        return 'Equivalence';
      case QuestionCategory.inference:
        return 'Inference';
    }
  }

  static String getDisplayNameFromName(String name) {
    try {
      final category = QuestionCategory.values.firstWhere((e) => e.name == name);
      return category.displayName;
    } catch (e) {
      return name; // 如果找不到匹配的枚举值，返回原始名称
    }
  }

  static String getNameFromDisplayName(String displayName) {
    try {
      final category = QuestionCategory.values.firstWhere((e) => e.displayName == displayName);
      return category.name;
    } catch (e) {
      return displayName; // 如果找不到匹配的枚举值，返回原始显示名称
    }
  }
}

enum QuestionDifficulty {
  easy,
  medium,
  hard,
}

extension QuestionDifficultyExtension on QuestionDifficulty {
  String get displayName {
    switch (this) {
      case QuestionDifficulty.easy:
        return 'Easy';
      case QuestionDifficulty.medium:
        return 'Medium';
      case QuestionDifficulty.hard:
        return 'Hard';
    }
  }

  static String getDisplayNameFromName(String name) {
    try {
      final difficulty = QuestionDifficulty.values.firstWhere((e) => e.name == name);
      return difficulty.displayName;
    } catch (e) {
      return name; // 如果找不到匹配的枚举值，返回原始名称
    }
  }

  static String getNameFromDisplayName(String displayName) {
    try {
      final difficulty = QuestionDifficulty.values.firstWhere((e) => e.displayName == displayName);
      return difficulty.name;
    } catch (e) {
      return displayName; // 如果找不到匹配的枚举值，返回原始显示名称
    }
  }
}
